### DISCUSSION TITLE
Matrix Keypad

### DISCUSSION THREAD

**Discussion by Oekel** [arendst/Tasmota#22598] [Q&A] [answered]
Hi,

I have a default 3x4 Keypad that is working on a Matrix of rows and colums.
So with a 100% custom firmware the task is clear: Giving power on the first of the rows as output-pin (timemultiplexed) and check all colums as input-pins. If there is a "shortcut" the responsible key is pressed on keypad. (Working the same if switching rows and colums as input/output)

But I am using THIS ESP allready for other task and I like tasmota.

So how do I configure this timemultiplexing and query with rules or whatever is available on tasmota?
At the end I want to write ALL of these 12 characters to the same MQTT-Topic, so that I can work "onChange" in my broker.

Can someone give me an example? Or Maybee even full code for 3x3 ---> "1,2,3"; "4,5,6"; "7,8,9"
(3x4 += "*,0,#")

Whould be great to get any hint, becourse most of the time I use Tasmota as given (max compiled with custom flags but no custom scrips or rules)

Nice Holidays Oekel

**Comment by barbudor** [comment]
Scanning a key matrix requires quite a fast processing to be efficient.
So yes you could define 3 GPIOs as Relay and 4 as switches. Using interlocked relays will help to have only 1 ON at a time. Then rules timer to schedule switching on one after the other
Then using rules to catch the switches and converting in a key number may be possible.
But that will be slow.

If you have an ESP32, you may try Berry that should make it easier

I fear that you would need a native driver to achieve correct behavior

Alternatively you can use a side components to manage the keypad like an standard arduino which would communicate with Tasmota through a serial

**Comment by Oekel** [comment, marked as answer]
The best solution would be supporting KeyScan with:
**TCA8418**
or
**HT16K33** (as DisplayMatrix allready in CodeBase)

So I created two Support-Request. (Hopefully I will join cod-writing) but for know I still hope for help of people who are allready in code ;)
